Regression in net/ethernet.h: C++ build failure (invalid cast from const void *)

### Describe the bug

Building Zephyr public headers from a C++ project (for example, Matter, which consumes Zephyr headers but is built outside Zephyr) fails due to an invalid pointer conversion error.

This issue appears to be a regression introduced by PR #105846
The same code compiled cleanly from C++ before this change.

**Error message:**
`error: invalid conversion from 'const void *' to 'const struct ethernet_api *' [-fpermissive]`

**Relevant context from the compiler output:**
```
In file included from zephyr/include/zephyr/net/wifi_mgmt.h:19,
                 from .../WiFiManager.h:32,
                 from .../DiagnosticDataProviderImplWiFi.cpp:25:

zephyr/include/zephyr/net/ethernet.h:949:47: error:
invalid conversion from "const void *" to "const struct ethernet_api *"
```
**Failing line in include/zephyr/net/ethernet.h** [L949](https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/net/ethernet.h#L949):
```
const struct ethernet_api *eth = dev->api;
```

**Expected behavior**:
Zephyr public headers should compile cleanly when included from C++ code, as they did before.

**Actual behavior**:
C++ compilation fails due to an implicit cast from `const void * (dev->api)` to `const struct ethernet_api *`.

**Suggested fix**
Applying an explicit cast fixes the issue locally:
```
const struct ethernet_api *eth = (const struct ethernet_api *) dev->api;
```


### Regression

- [x] This is a regression.

### Steps to reproduce

- Build or include Zephyr headers from a C++ project (e.g. Matter or another external consumer).
- Ensure include/zephyr/net/ethernet.h is pulled in (e.g. via wifi_mgmt.h).
- Compile a .cpp file with a standard C++ toolchain.

### Relevant log output

```shell

```

### Impact

Major – Severely degrades functionality; workaround is difficult or unavailable.

### Environment

- Zephyr: main (after PR #105846)
- Language mode: C++
- Compiler: GCC (C++)
- Build type: external C++ project including Zephyr public headers (e.g. Matter)

### Additional Context

_No response_

--- a/include/zephyr/net/ethernet.h
+++ b/include/zephyr/net/ethernet.h
@@ -946,7 +946,7 @@ int net_eth_get_hw_config(struct net_if *iface, enum ethernet_config_type type,
 			 struct ethernet_config *config)
 {
 	const struct device *dev = net_if_get_device(iface);
-	const struct ethernet_api *eth = dev->api;
+	const struct ethernet_api *eth = (struct ethernet_api *)dev->api;
 
 	if (!eth->get_config) {
 		return -ENOTSUP;

